src/commands/data/schemas/util/syncSchema.ts

```diff
@@ -359,3 +297,185 @@ function deepDiff(object: any, other: any) {
+}) {
+  console.group(group);
+  if (
+    toAdd.length < 1 &&
+    toRemove.length < 1 &&
+    toUpdate.length < 1
+  ) {
+    console.log('No update required');
+  }
+  for (const key of toAdd) {
+    console.log(`Will be added: ${chalk.green(getIdentifier(key))}`);
+  }
+  for (const key of toRemove) {
+    console.log(`Will be removed: ${chalk.red(getIdentifier(key))}`);
+  }
+  for (const key of toUpdate) {
+    console.log(`Will be updated: ${chalk.yellow(getIdentifier(key))}`);
+  }
+  console.groupEnd();
+}
+
+function logChangesInIndexes(schema: any, { newIndexes, removedIndexes }: {
+  newIndexes: any[];
+  removedIndexes: any[];
+}) {
+  const changes = compareArraysByName(newIndexes, removedIndexes);
+  logChanges(`Schema ${schema.name} - Indexes`, {
+    toAdd: changes.toAdd.map(v => v.name),
+    toRemove: changes.toRemove.map(v => v.name),
+    toUpdate: changes.toUpdate.map(v => v.name),
+  });
+}
+
+export function compareSchemaKey(localSchema: any, cloudSchema: any, key: string) {
+  if (!localSchema[key] && !cloudSchema[key]) {
+    return {
+      toAdd: [],
+      toRemove: [],
+      toUpdate: [],
+    };
+  }
+
+  if (
+    Array.isArray(localSchema[key]) ||
+    (!localSchema[key] && Array.isArray(cloudSchema[key]))
+  ) {
+    return compareArraysByName(
+      localSchema[key],
+      cloudSchema[key]
+    );
+  }
+  return compareSchemas(
+    localSchema[key],
+    cloudSchema[key]
+  );
+}
+
+function compareSchemas(localSchema: any, cloudSchema: any) {
+  const toAdd = _.difference(
+    Object.keys(localSchema),
+    Object.keys(cloudSchema)
+  );
+
+  // calculate excess properties
+  const toRemove = _.difference(
+    Object.keys(cloudSchema),
+    Object.keys(localSchema)
+  );
+
+  // calculate properties which have changed
+  const toUpdate: any = Object.keys(deepDiff(
+    _(localSchema).omit(toAdd).value(),
+    _(cloudSchema).omit(toRemove).value()
+  ));
+
+  return {
+    toAdd,
+    toRemove,
+    toUpdate,
+  };
+}
+
+function compareArraysByName(localSchema: any[], cloudSchema: any[]) {
+  const toAdd: any[] = _.differenceBy(
+    localSchema,
+    cloudSchema,
+    'name'
+  );
+
+  // evaluate excess
+  const toRemove: any[] = _.differenceBy(
+    cloudSchema,
+    localSchema,
+    'name'
+  );
+
+  // evaluate incorrect
+  const toUpdate: any = _.differenceWith(
+    // local schema, omit missing
+    _.differenceBy(localSchema, toAdd, 'name'),
+    // cloud schema, omit excess
+    _.differenceBy(cloudSchema, toRemove, 'name'),
+    // compare via helper function
+    (targetTransition, currentTransition) => _.isEqual(
+      _.omit(targetTransition, ['id']),
+      _.omit(currentTransition, ['id'])
+    )
+  );
+
+  return {
+    toAdd,
+    toRemove,
+    toUpdate,
+  };
+}
+
+function findTransitionByName(name: string, transitions: any[]) {
+  return _.find(
+    transitions,
+    transition => transition.name === name
+  );
+}
+
+function compareIndexes(localSchema: any, cloudSchema: any) {
```

Comment:
Twould also be nice if you liked some ideas in my readability examples if you could apply that here, but again it's opinion based.